pkg/fmt/alias.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fmt
+
+import (
+	"fmt"
+)
+
+type State interface {
+	fmt.State
+}
+
+type Formatter interface {
+	fmt.Formatter
+}
+
+type Stringer interface {
+	fmt.Stringer
+}
+
+type GoStringer interface {
+	fmt.GoStringer
+}
+
+var (
```

Comment:
Not sure how much value does these changes create. Just save typing `fmt.`?

Reply:
By this change you mean the whole PR? Exporting `fmt` funtions? Will answer both:

In case you refer to exporting `fmt`, if you need the new functions provided by this package, you import `"kubebuilder/pkg/fmt"` and it replaces your previous `"fmt"` import. This way you don't have to be providing import alias to packages. You would still need to write it `fmt.Xxxxxx()` but now you will also have `fmt.Printm` family.

In case you refer to the whole PR, this does add functionality. This formatting (`"%[kind]s" getting replaced by `strings.ToLower(resource.Kind)`) does not only apply to default paths. Paths provided outside the `GetInput` function do also benefit from it. This allows to provide paths **before** having access to the resource like in a flag (for implementing a flag that is in `kubebuilder` but breaks the scaffolding right now, the boilerplate path.

Reply:
IHMO: 

We should not move forward with this change. My vote for this is -1  since it makes the maintenance ability be lower and do not improve the performance and/or design either. 

Following some reasons.

- The fmt is part of the golang lib and everybody knows how to use and I do see any problem with write `fmt.Xxxxxx()`  or read/understand `fmt.Xxxxxx()`
- Provide `"kubebuilder/pkg/fmt"` would allow people to import it and create a need to support a new fmt lib as well
- Replace the get/set done directly in the attribute (`r.Resource.Attribute`) for a map["entry"] not implicit yet as `"%[kind]s"` make harder understanding from where the information came and how it works.  
- It will increase unnecessarily the usage of resources. 
- Regards:
   
> In case you refer to the whole PR, this does add functionality. This formatting ("%[kind]s" getting replaced by strings.ToLower(resource.Kind))

It can be solved in other ways as for example: Add a new attribute to Resources struct such as e.g `LowerKind` as it is done for other similar scenarios.

Reply:
> * The fmt is part of the golang and everybody knows how to use and I do see any problem with write `fmt.Xxxxxx()`  or read/understand `fmt.Xxxxxx()`

As I said before you still use `fmt` the same way. You are not removing the `fmt.Xxxxx` syntax.
<details><summary>Example usage</summary>

Standard `fmt` package:
```
package main

import (
    "fmt"
)

func main() {
    fmt.Println("Hello world!")
}
```

Extended `kubeuilder/pkg/fmt` package
```
package main

import (
    "fmt"
    fmt2 "kubebuilder/pkg/fmt" // We added this import
)

func main() {
    fmt.Println("Hello world!")
    // Now we can also use fmt2.Printm(..., map[string]string{...})
}
```

Extended `kubeuilder/pkg/fmt` package that provides alias to standard `fmt`
```
package main

import (
    "kubebuilder/pkg/fmt" // Instead of adding an import we changed it
)

func main() {
    fmt.Println("Hello world!") // This line is exactly the same
    // Now we can also use fmt.Printm(..., map[string]string{...})
}
```

</details>

As you can see, providing the alias to the standard `fmt` package just makes usage exactly the same as if it was the standard package.

> * Provide `"kubebuilder/pkg/fmt"` would allow people to import it and create a need to support a new fmt lib as well

This can be solved by moving it to `kubebuilder/pkg/internal/fmt`, which will prevent external usage.

> * Replace the get/set done directly in the attribute (`r.Resource.Attribute`) for a `map["entry"]` not implicit yet as `"%[kind]s"` make harder understanding from where the information came and how it works.

But it allows us to define paths in flags, or other places where we can just write a string but we don't have the resource available. Basically it offers templating capabilities to paths the same way that files can template parts with the `{{ .Xxxxx }}` syntax. One example of a path that is suposed to be provided as a string is with the `kubebuilder init --flag ...` which actually doesn't work due to bug #1150.

> * It will increase unnecessarily the usage of resources.

You are saying that it will be less performant? This is not a performance PR, it is adding a functionality and the performance difference will be negligible. We are templating whole files, simple templating for path will not have a visible effect.

> > In case you refer to the whole PR, this does add functionality. This formatting ("%[kind]s" getting replaced by strings.ToLower(resource.Kind))
> 
> It can be solved in other ways as for example: Add a new attribute to Resources struct such as e.g `LowerKind` as it is done for other similar scenarios.

The problem is not the strings.ToLower part. The example still aplies to other replacements. Please re-read that sentence after I edit the comment to a example that doesn't include `strings.ToLower`. The point is not that it includes the lowercasing part, the point is that the templating is done to non-default paths too, where we may not have the resource variable yet, and then cannot use it to get the path.